core/src/main/java/com/google/cloud/sql/core/CoreSocketFactory.java

```diff
@@ -230,24 +206,28 @@ private boolean runningOnGaeStandard() {
   // TODO(berezv): separate creating socket and performing connection to make it easier to test
   @VisibleForTesting
   Socket createSslSocket(String instanceName, List<String> ipTypes) throws IOException {
+    CloudSqlInstance instance =
+        instances.computeIfAbsent(
+            instanceName, k -> new CloudSqlInstance(k, adminApi, executor, localKeyPair));
+
     try {
-      return createAndConfigureSocket(instanceName, ipTypes, CertificateCaching.USE_CACHE);
-    } catch (SSLHandshakeException err) {
-      logger.warning(
-          String.format(
-              "SSL handshake failed for Cloud SQL instance [%s], "
-                  + "retrying with new certificate.\n%s",
-              instanceName, Throwables.getStackTraceAsString(err)));
-
-      if (!forcedRenewRateLimiter.tryAcquire()) {
-        logger.warning(
-            String.format(
-                "Renewing too often, rate limiting certificate renewal for Cloud SQL "
-                    + "instance [%s].",
-                instanceName));
-        forcedRenewRateLimiter.acquire();
-      }
-      return createAndConfigureSocket(instanceName, ipTypes, CertificateCaching.BYPASS_CACHE);
+      SSLSocket socket = instance.createSslSocket();
+
+      // TODO(kvg): Support all socket related options listed here:
+      // https://dev.mysql.com/doc/connector-j/en/connector-j-reference-configuration-properties.html
+      socket.setKeepAlive(true);
+      socket.setTcpNoDelay(true);
+
+      String instanceIp = instance.getPreferredIp(ipTypes);
+
+      socket.connect(new InetSocketAddress(instanceIp, serverProxyPort));
+      socket.startHandshake();
+
+      return socket;
+    } catch (Exception ex) {
```

Comment:
TODO for later: Catch `IOException` only, and move the `forceRefresh()` call into `getInstanceData()`.